apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/TableSchemaExtensions.cs

```diff
@@ -43,33 +43,52 @@ internal static class TableSchemaExtensions
 
         internal static string BuildSelectedFields(this TableSchema schema)
         {
-            if (!(schema?.Fields?.Count > 0))
+            // Althouhg null and empty SelectedFields mean the same, that is to return
+            // whole rows, let's return null for a null schema and string.Empty for an
+            // empty schema, just in case these get diferentiated in the future. And also
+            // it's clearer.
+            if (schema == null)
+            {
+                return null;
+            }
+            if (!(schema.Fields?.Count > 0))
             {
                 return string.Empty;
             }
 
+            // This list will contain the paths to all terminal fields that we have found.
+            // In the end this list will contain all selected fields.
             var built = new List<string>();
-            var currentField = new List<string>();
-            BuildSelectedFields(schema.Fields, currentField, built);
+            // This list will contain the path to the field we are currently traversing.
+            // If it is a terminal field (no nested fields) then we build the path to it
+            // and backtrack to the previous level.
+            // If it is not terminal, then we just traverse (in depth) its nested fields.
+            var currentPath = new List<string>();
+            BuildSelectedFields(schema.Fields, currentPath, built);
             return string.Join(",", built);
         }
 
-        private static void BuildSelectedFields(IList<TableFieldSchema> fieldSchemas, IList<string> currentField, IList<string> built)
+        private static void BuildSelectedFields(IList<TableFieldSchema> fieldSchemas, IList<string> currentPath, IList<string> built)
         {
-            // Whatever we have in currentField is terminal.
-            // Build that an we are done.
             if (!(fieldSchemas?.Count > 0))
             {
-                built.Add(string.Join(".", currentField));
+                // Whatever we have in currentPath is terminal.
+                // Build that and we are done.
+                built.Add(string.Join(".", currentPath));
             }
             else
             {
-                int removeAtIndex = currentField.Count;
-                foreach(var field in fieldSchemas)
+                // For each field at this level, we'll add it to the current path
+                // and then go in depth into its possible nested fields.
+                // When we return from the recursive call, we can remove this field
+                // from the current path. It will always be at the last index, which
+                // will always be one more than what we have right now.
```

Comment:
It's still not clear to me why we can't just remove the last entry *before* adding the child entries. (That would be very slightly more efficient, but that's not particularly important.)

Reply:
Because the one I'm removing is the parent of those child entries, it is part of the path to get to each of the terminal child ones. I add it inside the loop, before doing the recursive call and remove it right after returning. I might be missing something though?

Reply:
Ah, okay, I think I see now. It might be a bit clearer with method documentation for what the various parameters mean. Maybe we should merge as is, then I should spend a bit more time understanding what this does, then propose documentation? (It's only a private method, so it's not urgent.)

Reply:
Okay, this is just like an indepth traversal of a forest, only I dind't use a stack, which would probably have made it clearer, because when I get to a terminal field I want to build the path wich is FIFO and not LIFO, I could have reversed the stack etc., but with the list I get both easier, FIFO for building the path, LIFO for backtracking.
I'll merge now, but happy to go over this later for improvements and docs.